src/storage/src/dedup_pk_cell_based_row_deserializer.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::Deref;
+
+use risingwave_common::array::Row;
+use risingwave_common::catalog::OrderedColumnDesc;
+use risingwave_common::error::Result;
+use risingwave_common::types::{Datum, VirtualNode};
+use risingwave_common::util::ordered::OrderedRowDeserializer;
+
+use crate::cell_based_row_deserializer::{CellBasedRowDeserializer, ColumnDescMapping};
+
+/// Similar to [`CellBasedRowDeserializer`], but for dedup pk cell encoding.
+#[derive(Clone)]
+pub struct DedupPkCellBasedRowDeserializer<Desc: Deref<Target = ColumnDescMapping>> {
+    pk_deserializer: OrderedRowDeserializer,
+    inner: CellBasedRowDeserializer<Desc>,
+
+    // Maps pk fields with:
+    // 1. same value and memcomparable encoding,
+    // 2. corresponding row positions.
+    pk_to_row_mapping: Vec<Option<usize>>,
+}
+
+impl<Desc: Deref<Target = ColumnDescMapping>> DedupPkCellBasedRowDeserializer<Desc> {
+    /// Create a [`DedupPkCellBasedRowDeserializer`]
+    /// to decode cell based row with dedup pk encoding.
+    pub fn new(column_mapping: Desc, pk_descs: &[OrderedColumnDesc]) -> Self {
+        let (pk_data_types, pk_order_types) = pk_descs
```

Comment:
The RowDeserializer / Serializer was designed to be created without overhead, and that's why I introduced `ColumnDescMapping` before to cache the HashMap of column mapping. For DedupPk serializer, it would be better to have a new `DedupColumnDescMapping` to store all such generated info (e.g. pk_to_row_mapping). Can done in later PRs.

Reply:
The ultimate solution will be to cache all such information in a `StateTableInfo` struct, and pass it everywhere... I guess @.BugenZhao is working on this.